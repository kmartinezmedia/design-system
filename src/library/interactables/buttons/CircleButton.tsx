import React, { FunctionComponent } from 'react';
import {
  TouchableWithoutFeedback,
  GestureResponderEvent,
  Animated,
  StyleSheet,
} from 'react-native';
import { View, ViewProps } from '@designSystem/primitives';
import { usePressAnimation, useButtonVariant } from '@designSystem/hooks';
import { sizing } from '@designSystem/theme';

export interface CircleButtonProps extends Omit<ViewProps, 'onPress'> {
  disabled?: boolean;
  variant?: 'primary' | 'secondary';
  onPress?: (event: GestureResponderEvent) => void;
  testID?: string;
}

export const CircleButton: FunctionComponent<CircleButtonProps> = ({
  children,
  disabled,
  onPress,
  testID,
  variant = 'primary',
  ...props
}) => {
  const [pressIn, pressOut, scale] = usePressAnimation();
  const variantStyles = useButtonVariant(variant);
  return (
    <View {...props}>
      <TouchableWithoutFeedback
        onPress={onPress}
        onPressIn={pressIn}
        onPressOut={pressOut}
        disabled={disabled}>
        <Animated.View
          style={[
            styles.circleButton,
            disabled && styles.disabled,
            { transform: [{ scale }] },
            variantStyles,
          ]}
          testID={testID}>
          {children}
        </Animated.View>
      </TouchableWithoutFeedback>
    </View>
  );
};

const styles = StyleSheet.create({
  circleButton: {
    height: sizing.icons.button,
    width: sizing.icons.button,
    borderRadius: sizing.icons.button / 2,
    flexDirection: 'column',
    justifyContent: 'center',
    alignItems: 'center',
  },
  disabled: {
    opacity: 0.5,
  },
});
