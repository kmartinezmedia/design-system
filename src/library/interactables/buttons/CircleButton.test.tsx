import React from 'react';
import { Text } from 'react-native';
import { create, act, ReactTestRenderer } from 'react-test-renderer';
import { CircleButton } from './CircleButton';

describe('CircleButton', () => {
  it('default', () => {
    let component: ReactTestRenderer;
    act(() => {
      component = create(
        <>
          <CircleButton>
            <Text>🐛</Text>
          </CircleButton>
          <CircleButton variant="secondary">
            <Text>🐛</Text>
          </CircleButton>
          <CircleButton disabled>
            <Text>🐛</Text>
          </CircleButton>
        </>,
      );
    });
    expect(component!.toJSON()).toMatchSnapshot();
  });
  it('positions', () => {
    let component: ReactTestRenderer;
    act(() => {
      component = create(
        <>
          <CircleButton spacing={3}>
            <Text>🐛</Text>
          </CircleButton>
          <CircleButton spacing={2} pin="right">
            <Text>🐛</Text>
          </CircleButton>
        </>,
      );
    });
    expect(component!.toJSON()).toMatchSnapshot();
  });
});
