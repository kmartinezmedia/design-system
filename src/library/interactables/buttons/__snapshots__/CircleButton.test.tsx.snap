// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CircleButton default 1`] = `
Array [
  <View
    style={
      Array [
        Object {
          "alignItems": "stretch",
          "alignSelf": "auto",
          "flexDirection": "column",
          "flexGrow": undefined,
          "flexShrink": undefined,
          "flexWrap": "nowrap",
          "height": undefined,
          "justifyContent": "flex-start",
          "width": undefined,
        },
        undefined,
        undefined,
        undefined,
        Object {},
      ]
    }
  >
    <TouchableWithoutFeedback
      onPressIn={[Function]}
      onPressOut={[Function]}
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#1652F0",
            "borderColor": "#1652F0",
            "borderRadius": 20,
            "borderWidth": 1,
            "flexDirection": "column",
            "height": 40,
            "justifyContent": "center",
            "textColor": "#FFFFFF",
            "transform": Array [
              Object {
                "scale": 1,
              },
            ],
            "width": 40,
          }
        }
      >
        <Text>
          🐛
        </Text>
      </View>
    </TouchableWithoutFeedback>
  </View>,
  <View
    style={
      Array [
        Object {
          "alignItems": "stretch",
          "alignSelf": "auto",
          "flexDirection": "column",
          "flexGrow": undefined,
          "flexShrink": undefined,
          "flexWrap": "nowrap",
          "height": undefined,
          "justifyContent": "flex-start",
          "width": undefined,
        },
        undefined,
        undefined,
        undefined,
        Object {},
      ]
    }
  >
    <TouchableWithoutFeedback
      onPressIn={[Function]}
      onPressOut={[Function]}
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#FFFFFF",
            "borderColor": "#F1F3F5",
            "borderRadius": 20,
            "borderWidth": 1,
            "flexDirection": "column",
            "height": 40,
            "justifyContent": "center",
            "textColor": "#1652F0",
            "transform": Array [
              Object {
                "scale": 1,
              },
            ],
            "width": 40,
          }
        }
      >
        <Text>
          🐛
        </Text>
      </View>
    </TouchableWithoutFeedback>
  </View>,
  <View
    style={
      Array [
        Object {
          "alignItems": "stretch",
          "alignSelf": "auto",
          "flexDirection": "column",
          "flexGrow": undefined,
          "flexShrink": undefined,
          "flexWrap": "nowrap",
          "height": undefined,
          "justifyContent": "flex-start",
          "width": undefined,
        },
        undefined,
        undefined,
        undefined,
        Object {},
      ]
    }
  >
    <TouchableWithoutFeedback
      disabled={true}
      onPressIn={[Function]}
      onPressOut={[Function]}
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#1652F0",
            "borderColor": "#1652F0",
            "borderRadius": 20,
            "borderWidth": 1,
            "flexDirection": "column",
            "height": 40,
            "justifyContent": "center",
            "opacity": 0.5,
            "textColor": "#FFFFFF",
            "transform": Array [
              Object {
                "scale": 1,
              },
            ],
            "width": 40,
          }
        }
      >
        <Text>
          🐛
        </Text>
      </View>
    </TouchableWithoutFeedback>
  </View>,
]
`;

exports[`CircleButton positions 1`] = `
Array [
  <View
    style={
      Array [
        Object {
          "alignItems": "stretch",
          "alignSelf": "auto",
          "flexDirection": "column",
          "flexGrow": undefined,
          "flexShrink": undefined,
          "flexWrap": "nowrap",
          "height": undefined,
          "justifyContent": "flex-start",
          "width": undefined,
        },
        Object {
          "padding": 12,
        },
        undefined,
        undefined,
        Object {},
      ]
    }
  >
    <TouchableWithoutFeedback
      onPressIn={[Function]}
      onPressOut={[Function]}
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#1652F0",
            "borderColor": "#1652F0",
            "borderRadius": 20,
            "borderWidth": 1,
            "flexDirection": "column",
            "height": 40,
            "justifyContent": "center",
            "textColor": "#FFFFFF",
            "transform": Array [
              Object {
                "scale": 1,
              },
            ],
            "width": 40,
          }
        }
      >
        <Text>
          🐛
        </Text>
      </View>
    </TouchableWithoutFeedback>
  </View>,
  <View
    style={
      Array [
        Object {
          "alignItems": "stretch",
          "alignSelf": "auto",
          "flexDirection": "column",
          "flexGrow": undefined,
          "flexShrink": undefined,
          "flexWrap": "nowrap",
          "height": undefined,
          "justifyContent": "flex-start",
          "width": undefined,
        },
        Object {
          "padding": 8,
        },
        Object {
          "bottom": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        undefined,
        Object {},
      ]
    }
  >
    <TouchableWithoutFeedback
      onPressIn={[Function]}
      onPressOut={[Function]}
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#1652F0",
            "borderColor": "#1652F0",
            "borderRadius": 20,
            "borderWidth": 1,
            "flexDirection": "column",
            "height": 40,
            "justifyContent": "center",
            "textColor": "#FFFFFF",
            "transform": Array [
              Object {
                "scale": 1,
              },
            ],
            "width": 40,
          }
        }
      >
        <Text>
          🐛
        </Text>
      </View>
    </TouchableWithoutFeedback>
  </View>,
]
`;
